src/Libraries/DSCPython/DCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,806 @@
+        public Dictionary<string, Type> ImportedTypes { get; set; }
+
+        /// <summary>
+        /// Generate completion data for the specified text, while import the given types into the
+        /// scope and discovering variable assignments.
+        /// </summary>
+        /// <param name="code">The code to parse</param>
+        /// <param name="expand">Determines if the entire namespace should be used</param>
+        /// <returns>Return a list of IExternalCodeCompletionData </returns>
+        public IExternalCodeCompletionData[] GetCompletionData(string code, bool expand = false)
+        {
+            IEnumerable<PythonCodeCompletionDataCore> items = new List<PythonCodeCompletionDataCore>();
+
+            if (!PythonEngine.IsInitialized)
+            {
+                PythonEngine.Initialize();
+                PythonEngine.BeginAllowThreads();
+            }
+
+            IntPtr gs = PythonEngine.AcquireLock();
+            try
+            {
+                using (Py.GIL())
+                {
+                    if (code.Contains("\"\"\""))
+                    {
+                        code = PythonCodeCompletionUtils.StripDocStrings(code);
+                    }
+
+                    if (Scope == null)
+                    {
+                        Scope = CreateUniquePythonScope();
+                    }
+
+                    UpdateImportedTypes(code);
+                    UpdateVariableTypes(code);  // Possibly use hindley-milner in the future
+
+                    // If expand param is true use the entire namespace from the line of code
+                    // Else just return the last name of the namespace
+                    string name = expand ? PythonCodeCompletionUtils.GetLastNameSpace(code) : PythonCodeCompletionUtils.GetLastName(code);
+                    if (!String.IsNullOrEmpty(name))
+                    {
+                        try
+                        {
+                            // Attempt to get type using naming
+                            Type type = expand ? PythonCodeCompletionUtils.TryGetTypeFromFullName(name) : TryGetType(name);
+
+                            // CLR type
+                            if (type != null)
+                            {
+                                items = EnumerateMembers(type, name).Select(x => new PythonCodeCompletionDataCore(x.Item1, x.Item2, x.Item3, x.Item4, this));
+                            }
+                            // Variable type
+                            else if (VariableTypes.TryGetValue(name, out type))
+                            {
+                                items = EnumerateMembers(type, name).Select(x => new PythonCodeCompletionDataCore(x.Item1, x.Item2, x.Item3, x.Item4, this));
+                            }
+                            else
+                            {
+                                var mem = LookupMember(name);
+                                var pythonModule = mem as PyObject;
```

Comment:
why does this only cover modules? Wouldn't this cover any pyobject?

Reply:
renamed